File: lib/models/transaction.dart
import 'package:isar/isar.dart';

part 'transaction.g.dart';

/// 账单实体
@collection
class Transaction {
  Id id = Isar.autoIncrement;

  /// 消费内容，如"买面包"
  @Index()
  String description;

  /// 金额，如 5.0
  double amount;

  /// 类别：餐饮/交通/购物/娱乐/居住/医疗/教育/其他
  @Index()
  String category;

  /// 消费日期（归一化到当天 00:00:00）
  @Index()
  DateTime date;

  /// 记录创建时间，用于排序
  DateTime createdAt;

  Transaction({
    this.id = Isar.autoIncrement,
    required this.description,
    required this.amount,
    required this.category,
    required this.date,
    required this.createdAt,
  });

  Transaction copyWith({
    int? id,
    String? description,
    double? amount,
    String? category,
    DateTime? date,
    DateTime? createdAt,
  }) {
    return Transaction(
      id: id ?? this.id,
      description: description ?? this.description,
      amount: amount ?? this.amount,
      category: category ?? this.category,
      date: date ?? this.date,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

/// 预设类别常量
class Categories {
  static const List<String> all = [
    '餐饮',
    '交通',
    '购物',
    '娱乐',
    '居住',
    '医疗',
    '教育',
    '其他',
  ];

  static const String other = '其他';

  /// 校验类别合法性，非法值归入"其他"
  static String normalize(String? category) {
    if (category == null) return other;
    final trimmed = category.trim();
    return all.contains(trimmed) ? trimmed : other;
  }
}

File: voxspend-web/src/models/transaction.ts
export interface Transaction {
  id?: number
  description: string
  amount: number
  category: string
  date: string // YYYY-MM-DD
  createdAt: number // timestamp
}

export const CATEGORIES = ['餐饮', '交通', '购物', '娱乐', '居住', '医疗', '教育', '其他'] as const
export type Category = (typeof CATEGORIES)[number]

export const CATEGORY_COLORS: Record<string, string> = {
  '餐饮': '#FF9500',
  '交通': '#5856D6',
  '购物': '#FF2D55',
  '娱乐': '#AF52DE',
  '居住': '#34C759',
  '医疗': '#FF3B30',
  '教育': '#5AC8FA',
  '其他': '#8E8E93',
}

export function normalizeCategory(category: string | null | undefined): string {
  if (!category) return '其他'
  const trimmed = category.trim()
  return (CATEGORIES as readonly string[]).includes(trimmed) ? trimmed : '其他'
}

export function normalizeDate(date: Date): string {
  return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}-${String(date.getDate()).padStart(2, '0')}`
}
